protostar/cairo/cairo_bindings.py

```diff
@@ -17,41 +20,55 @@ def compile_starknet_contract(
-    return cairo_python_bindings.compile_starknet_contract_from_path(  # pyright: ignore
-        str(input_path),
-        str(output_path) if output_path else None,
-        [str(path) for path in cairo_path] if cairo_path else None,
-    )
+    with handle_bindings_errors("compile_starknet_contract"):
+        return cairo_python_bindings.compile_starknet_contract_from_path(  # pyright: ignore
+            str(input_path),
+            str(output_path) if output_path else None,
+            [str(path) for path in cairo_path] if cairo_path else None,
+        )
 
 
 def collect_tests(
     input_path: Path,
     output_path: Optional[Path] = None,
     cairo_path: Optional[list[Path]] = None,
 ) -> TestCollectorOutput:
-    output = cairo_python_bindings.collect_tests(  # pyright: ignore
-        str(input_path),
-        str(output_path) if output_path else None,
-        [str(path) for path in cairo_path] if cairo_path else None,
-    )
-    return TestCollectorOutput(sierra_output=output[0], test_names=output[1])
+    with handle_bindings_errors("collect_tests"):
+        output = cairo_python_bindings.collect_tests(  # pyright: ignore
+            str(input_path),
+            str(output_path) if output_path else None,
+            [str(path) for path in cairo_path] if cairo_path else None,
+        )
+        return TestCollectorOutput(sierra_output=output[0], test_names=output[1])
 
 
 def compile_protostar_sierra_to_casm_from_path(
     named_tests: list[str], input_path: Path, output_path: Optional[Path] = None
 ) -> Optional[dict]:
-    compiled_str = cairo_python_bindings.compile_protostar_sierra_to_casm_from_path(  # pyright: ignore
-        named_tests, str(input_path), str(output_path) if output_path else None
-    )
-    return json.loads(compiled_str)
+    with handle_bindings_errors("compile_protostar_sierra_to_casm_from_path"):
+        compiled_str = cairo_python_bindings.compile_protostar_sierra_to_casm_from_path(  # pyright: ignore
+            named_tests, str(input_path), str(output_path) if output_path else None
+        )
+        return json.loads(compiled_str)
 
 
 def compile_protostar_sierra_to_casm(
     named_tests: list[str], input_data: str, output_path: Optional[Path] = None
 ) -> Optional[dict]:
-    compiled_str = (
-        cairo_python_bindings.compile_protostar_sierra_to_casm(  # pyright: ignore
-            named_tests, input_data, str(output_path) if output_path else None
+    with handle_bindings_errors("compile_protostar_sierra_to_casm"):
+        compiled_str = (
+            cairo_python_bindings.compile_protostar_sierra_to_casm(  # pyright: ignore
+                named_tests, input_data, str(output_path) if output_path else None
+            )
         )
-    )
-    return json.loads(compiled_str)
+        return json.loads(compiled_str)
+
+
+@contextmanager
+def handle_bindings_errors(binding_name: str):
+    try:
+        yield
+    except RuntimeError as ex:
+        raise ProtostarException(
```

Comment:
fundamentally i think this is a hard topic because i think those should be distinguishable between user error and implementation error itself. Would it be possible to throw different exception types from the compiler? Or make the interface not throw at all, and communicate through return values only (dictionaries or something like that)?

Reply:
cc @MaksymilianDemitraszek

Reply:
Yup, ProtostarError should never be catched

Reply:
Ok, fair enough, I'd go a step further and name it `CairoBindingException` - this way all the errors that come from bindings will be unified and bindings don't only do compilation but also test collection for example (they may do different things in the future as well). If we ever need to distinguish the errors in the scope of bindings, we can always introduce a structure based on the `CairoBindingException`.

Reply:
[7050aec](https://github.com/software-mansion/protostar/pull/1553/commits/7050aec53aee3a08eea114cd3271b81d10873ce6)